Library/OcCryptoLib/BigNumLib.h

```diff
@@ -0,0 +1,126 @@
+/**
+  This library performs arbitrary precision arithmetic operations.
+  For more details, please refer to the source files and function headers.
+
+  Copyright (C) 2019, Download-Fritz. All rights reserved.
+
+This program and the accompanying materials
+are licensed and made available under the terms and conditions of the BSD License
+which accompanies this distribution.  The full text of the license may be found at
+http://opensource.org/licenses/bsd-license.php
+
+THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#ifndef BIG_NUM_LIB_H
+#define BIG_NUM_LIB_H
+
+//
+// Primitives
+//
+
+/**
+  Assigns A the value 0.
+
+  @param[in,out] A  The number to assign to.
+
+**/
+VOID
+BigNumAssign0 (
+  IN OUT OC_BN  *A
+  );
+
+/**
+  Parses a data array into a number. The buffer size must be a multiple of the
+  Word size. The length of Result must precisely fit the required size.
+
+  @param[in,out] Result      The buffer to store the result in.
+  @param[in]     Buffer      The buffer to parse.
+  @param[in]     BufferSize  The size, in bytes, of Buffer.
+  @param[in]     NumWords    The number of Words of Result.
+
+**/
+VOID
+BigNumDataParseBuffer (
+  IN OUT OC_BN_WORD       *Result,
```

Comment:
The interface seems counterintuitive to me:
- In the first place why do not we pass a pair of `OC_BN_WORD/OC_BN_NUM_WORDS` in all interfaces or use `OC_BN` in all interfaces? This interface feels like it is something internal, but exposed for some ugly reason.
- In addition to that, I would expect the parameter order to be `Result`, `ResultSize`, `Buffer`, `BufferSize`. Putting `NumWords` to the end and naming it like that makes things unreadable.

Reply:
I agree it is not intuitive, but it also is not supposed to be. Those "BigNumData(!)" functions are a thing so we can process BIGNUMs from the RSA files without re-allocating them as OC_BN. Except for the sub helper, they are all Montgomery functions, for which all operands and the destination always have the exact same size anyway, so it is safe.

Reply:
Well, apart from the argument order, which I believe should be changed, I can live with the rest. Yet I wonder whether we can actually make some hybrid format for this? Accept two arguments, and when Size is 0, read the first array entry as size… This should turn the struct into a union…

Reply:
For the argument order, I can live with either... struct discussion will be here: https://github.com/acidanthera/OcSupportPkg/pull/18#discussion_r341812347

Reply:
Well, I thought of unifying the interfaces, but it could have bad consequences indeed. Currently I do not like that we use `OC_BN *`, `OC_BN_WORD`, `OC_BN_WORD *`, `OC_BN_WORD *` + `Size`. Ideally I want one of that.

Reply:
`OC_BN_WORD` non-ptr is obviously independent of the rest (e.g. it is needed for the Montgomery Inverse as such) and `OC_BN_WORD *` without a size (e.g. NumWords in Montgomery) does not occur at all. It is basically just `OC_BN *` and `OC_BN_WORD *`, former has an explicit size (as part of the struct) and is used when sizes among operands may vary (Montgomery's R^2 mod N) and latter an implicit size and is used when all operands have the same size (Montgomery Multiplication). I'd like a unification too, but omiting the explicit size will make former harder to review, and enforcing explicit size will add overhead to RSA-file-based processing for no good reason as the additional safety is simply not needed (well, is not safety at all in this case). It's a tradeoff really, this is what I considered the least evil at the time.